### DISCUSSION TITLE
Mesh faces selection for applying BCs and Loads

### DISCUSSION THREAD

**Discussion by yvanblanchard** [adtzlr/felupe#837] [General]
Hello,
I would like to know if it would be possible to select/pick some faces (pick on 3D view of geometry, or select faces by Identifiers), using pyVista or other lib (NB: faces could be retrieved from mesh using 'edges detection').
Then the vertices (or elements faces) linked to those faces would be used to apply boundary conditions and/or loads, for doing analysis in Felupe.
Thank you !

**Comment by adtzlr** [comment]
Hi @yvanblanchard, in FElupe, there exists only a method to extract cell faces which are on the surface. You have to create a boundary region to access the attribute `RegionBoundary.mesh.cells_faces`. However, the surface normal vectors are already calculated at the quadrature points - by identifying the neighbors and  defining a threshold for the dot-products of normal vectors, we could split the surface in continuous feature faces. But there is no method available for that (yet). Would definitely be a good idea!

```python
import felupe as fem
import numpy as np
import pyvista as pv

radius = 1.0
mesh = fem.Circle(n=4, radius=radius)
region = fem.RegionQuadBoundary(mesh)
faces = region.mesh.cells_faces

assert np.allclose(mesh.x[faces] ** 2 + mesh.y[faces] ** 2, radius)

plotter = pv.Plotter()
actor = plotter.add_points(
    np.pad(mesh.points[np.unique(faces)], ((0, 0), (0, 1))),
    point_size=20,
    color="red",
)

mesh.plot(plotter=plotter, opacity=0.7).show()
```

![grafik](https://github.com/user-attachments/assets/cf41f95b-06ba-4553-8f93-834e73e1f63f)

Here you may find some more examples: https://felupe.readthedocs.io/en/latest/felupe/dof.html#felupe.Boundary

I'm no PyVista expert, so you'd have to ask your question in their repo how to do that in PyVista.

### See Also
- [`SolidBodyPressure`](https://felupe.readthedocs.io/en/latest/felupe/mechanics.html#felupe.SolidBodyPressure)
- [`PointLoad`](https://felupe.readthedocs.io/en/latest/felupe/mechanics.html#felupe.PointLoad)

**Comment by yvanblanchard** [comment]
Thank you !

Here is my part. Do you know if I can easily retrieve the regions (vertices or faces), using coordinates en length ?
Ex: green area (for load force) could be retrieved by filtering vertices of mesh with X coords range 0 to 10mm (size of left part) AND Z coords of greater or equal to 5mm (part thickness) AND  the Y coords (depth) of the range -12.5 to +12.5 mm.

![image](https://github.com/user-attachments/assets/15ac9a5a-29b5-46a9-a6d6-6d37dc689c99)

[PATTE.STL.txt](https://github.com/user-attachments/files/16801875/PATTE.STL.txt)

Thank you again.

**Comment by adtzlr** [comment]
To mask the points you mentioned:

```python
import felupe as fem
import numpy as np

mesh = fem.mesh.read("PATTE.STL")[0]

mask = np.logical_and.reduce(
    [mesh.x >= 0, mesh.x <= 10, mesh.z >= 5, mesh.y >= -12.5, mesh.y <= 12.5]
)
point_ids = np.arange(mesh.npoints)[mask]
```

However, there must be some problem with the model orientation or the units, because no points are selected. Could you please double-check your STL-file (or `meshio` has problems importing it)?

```python
plotter = mesh.plot()
plotter.show_grid()
plotter.show()
```

![grafik](https://github.com/user-attachments/assets/5b34d50d-0028-4087-aa3e-21347de3a834)

P.S.: Your mesh contains triangles of a solid body in 3D space. FElupe requires solid cells to work, e.g. triangles in 2D and tetrahedrons in 3D space. You must use some external library to create a tet-mesh from the surface triangles.

**Reply by yvanblanchard** [reply]
yes, the STL mesh file has meters units (instead of mm), and the orientation is not fine for capturing vertices by spatial coords, sorry.  I have to change this first...
But I have now the logical ; I just miss the part for applying BC & load on the 'point_Ids'.

I also have to fill the mesh skin with tets (after remeshing the skin with isotropic method), hopping pyVista could handle it, but I have doubt.

Thank you again Andreas!

**Reply by adtzlr** [reply]
You could have a look on [`scikit-gmsh`](https://github.com/pyvista/scikit-gmsh) but this package is in its early days. The mesh tools from `nschloe` (`meshpro`) are also great, some require a license. `nschloe/pygmsh` could do the job.

Back to FElupe: `mesh.x` is just a property which returns a read-only `mesh.points[:, 0]`.

**Reply by adtzlr** [reply]
You can pass the point mask to a displacement [boundary condition](https://felupe.readthedocs.io/en/latest/felupe/dof.html#felupe.Boundary).

```python
boundaries = dict(fixed=fem.Boundary(field, mask=mask))
```

A [`PointLoad`](https://felupe.readthedocs.io/en/latest/felupe/mechanics.html#felupe.PointLoad) requires the list of point ids.

```python
load = fem.PointLoad(field_container, points=point_ids, ...)
```
